### ISSUE TITLE
Client.shared.rx.connectionState remains disconnected after re-setting a user

### ISSUE THREAD

**Issue by vadubov** [GetStream/stream-chat-swift#267]
## What did you do?
I relogin my user by calling `Client.shared.disconnect()` and `Client.shared.set(user:tokenProvider:completion:)`, and then trying to load channels by `ChannelsPresenter` subscribing to `changes` and calling `reload()`.

## What did you expect to happen?
Channels loaded successfully.

## What happened instead?
No channels loaded, `changes` never fires.

## GetStream Environment
**GetStream Chat version:** 2.2.0
**iOS version:** 13.4
**Swift version:** 5
**Xcode version:** 11.4.1
**Device:** iPhone 11 Pro Simulator

## Additional context
After debugging it turns out that in completion of  `Client.shared.set(user:tokenProvider:completion:)` method `Client.shared.rx.connectionState` emits `.disconnected(nil)` even though `Client.shared.connectionState` is `.connected(UserConnection)`. Seems it prevents `Presenter.rx.prepareRequest` from starting a request.

**Comment by buh** [comment]
Thanks @vadubov!
I'll take a look!
Can you tell me do you use our `StreamChat` framework or only `StreamChatClient` and `StreamChatCore`?

**Comment by vadubov** [comment]
We use only `StreamChatClient` and `StreamChatCore`.

**Comment by buh** [comment]
Can you tell me more about your view hierarchy? 
Do you use `ChannelsViewController` in UITabBarController? If so, do you reload tabbar if your user do relogin?

**Comment by vadubov** [comment]
We use custom UI for channels and messages lists. We use UITabBarController, but reset UIWindow's rootViewController at logout. Also cancelling all subscription bags. Is it possible that the issue is caused by forgetting to cancel some subscription?

**Comment by buh** [comment]
Can you send me logs?

**Comment by vadubov** [comment]
Logging out...

`🐴 [15 May 11:04:52.149] Disconnecting deliberately...
🦄 [15 May 11:04:52.149] Clearing state after disconnect...
🦄 [15 May 11:04:52.149] Cancelling background work...
🦄 [15 May 11:04:52.149] Disconnecting: Resetting connection
connectionState 1
connectionState 2 state = disconnecting
2020-05-15 06:04:52 +0000: ChatManager.swift(112) setUser(_:apiTokenProvider:completion:):state = disconnecting

🦄 [15 May 11:04:52.166] Parsing WebSocket disconnect... (error: <nil>)
🦄 [15 May 11:04:52.166] Clearing state after disconnect...
2020-05-15 06:04:52 +0000: NotificationsCoordinator.swift(14) deinit:

🦄 [15 May 11:04:52.166] Cancelling background work...
2020-05-15 06:04:52 +0000: ChatsCoordinator.swift(15) deinit:

🦄 [15 May 11:04:52.166] 💔 Disconnected
2020-05-15 06:04:52 +0000: ProfileCoordinator.swift(16) deinit:

2020-05-15 06:04:52 +0000: FeedCoordinator.swift(14) deinit:

connectionState 1
connectionState 2 state = disconnected(nil)
2020-05-15 06:04:52 +0000: ChatManager.swift(112) setUser(_:apiTokenProvider:completion:):state = disconnected(nil)

🕸 State: unknown (-R -------)
🕸 Notifying stopped 🚶‍♂️`

Logging in...

`🦄 [15 May 11:05:29.497] Clearing state after disconnect...
🦄 [15 May 11:05:29.497] Cancelling background work...
🦄 [15 May 11:05:29.497] Skip disconnecting: WebSocket was not connected
connectionState 1
🐴 [15 May 11:05:29.498] 🀄️ Request for a new token from a token provider.
connectionState 2 state = disconnected(nil)

👤 Vit D1: 21
🀄️ Token: eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjEifQ.8cXsuIGDKlzthGgjlnbdvLYohDAVyLl1ExmC8A-jNqQ
🐴 [15 May 11:05:29.879] Request headers:
◾️ Authorization = eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjEifQ.8cXsuIGDKlzthGgjlnbdvLYohDAVyLl1ExmC8A-jNqQ
◾️ X-Stream-OS = iOS13.4.1
◾️ X-Stream-Client = stream-chat-swift-client-2.2.0
◾️ X-Stream-Device = x86_64
◾️ X-Stream-BundleId = com.saritasa.enterprise.witley.dev
◾️ Stream-Auth-Type = jwt
◾️ X-Stream-App-Environment = development

🕸 Notifying started 🏃‍♂️
🕸 State: available (-R -------)
🦄 [15 May 11:05:30.018] Cancelling background work...
🦄 [15 May 11:05:30.019] Connecting...
🦄 [15 May 11:05:30.019] ➡️ GET wss://chat-proxy-us-east.stream-io-api.com/connect?api_key=twj335e5wcah&authorization=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjEifQ.8cXsuIGDKlzthGgjlnbdvLYohDAVyLl1ExmC8A-jNqQ&stream-auth-type=jwt&json=%7B%22user_id%22:%2221%22,%22user_token%22:%22eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjEifQ.8cXsuIGDKlzthGgjlnbdvLYohDAVyLl1ExmC8A-jNqQ%22,%22server_determines_connection_id%22:true,%22user_details%22:%7B%22id%22:%2221%22,%22name%22:%22Vit%20D1%22,%22image%22:%22%5C/images%5C/college%5C/1%5C/6d378c06-5d98-4433-bc50-865e82aad6c1.jpeg%22%7D%7D
🦄 [15 May 11:05:30.019] URL query items:
▫️ authorization=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjEifQ.8cXsuIGDKlzthGgjlnbdvLYohDAVyLl1ExmC8A-jNqQ
▫️ stream-auth-type=jwt
▫️ json={
  "user_id" : "21",
  "user_token" : "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiMjEifQ.8cXsuIGDKlzthGgjlnbdvLYohDAVyLl1ExmC8A-jNqQ",
  "server_determines_connection_id" : true,
  "user_details" : {
    "id" : "21",
    "name" : "Vit D1",
    "image" : "\/images\/college\/1\/6d378c06-5d98-4433-bc50-865e82aad6c1.jpeg"
  }
}

🦄 [15 May 11:05:31.082] ❤️ Connected. Waiting for the current user data and connectionId...
🦄 [15 May 11:05:31.087] healthCheck 👤 21
🦄 [15 May 11:05:31.087] 📦  {
  "created_at" : "2020-05-15T06:05:30.911067609Z",
  "me" : {
    "id" : "21",
    "banned" : false,
    "roles" : [

    ],
    "unread_channels" : 2,
    "mutes" : [

    ],
    "last_active" : "2020-05-15T06:05:30.908461367Z",
    "created_at" : "2020-05-11T10:14:51.193855Z",
    "devices" : [

    ],
    "invisible" : false,
    "unread_count" : 9,
    "channel_mutes" : [

    ],
    "image" : "\/images\/college\/1\/6d378c06-5d98-4433-bc50-865e82aad6c1.jpeg",
    "updated_at" : "2020-05-15T05:54:51.470934Z",
    "role" : "user",
    "total_unread_count" : 9,
    "online" : true,
    "name" : "Vit D1"
  },
  "type" : "health.check",
  "connection_id" : "f707b5ac-5121-4b2c-921b-dbbf0f7f68ee",
  "cid" : "*"
}
🦄 [15 May 11:05:31.087] 🥰 Connected
2020-05-15 06:05:31 +0000: ChatManager.swift(109) setUser(_:apiTokenProvider:completion:):self.client.connectionState = connected(StreamChatClient.UserConnection(user: StreamChatClient.User(id: "21", created: 2020-05-11 10:14:51 +0000, updated: 2020-05-15 05:54:51 +0000, lastActiveDate: Optional(2020-05-15 06:05:30 +0000), isOnline: true, isInvisible: false, isBanned: false, role: StreamChatClient.User.Role.user, extraData: Optional(StreamChatClient.UserExtraData(name: Optional("Vit D1"), avatarURL: Optional(/images/college/1/6d378c06-5d98-4433-bc50-865e82aad6c1.jpeg))), devices: [], currentDevice: nil, mutedUsers: []), connectionId: "f707b5ac-5121-4b2c-921b-dbbf0f7f68ee"))

2020-05-15 06:05:31 +0000: ChatManager.swift(112) setUser(_:apiTokenProvider:completion:):state = disconnected(nil)
`
Notice the last two lines:

`2020-05-15 06:05:31 +0000: ChatManager.swift(109) setUser(_:apiTokenProvider:completion:):self.client.connectionState = connected(StreamChatClient.UserConnection(user: StreamChatClient.User(id: "21", created: 2020-05-11 10:14:51 +0000, updated: 2020-05-15 05:54:51 +0000, lastActiveDate: Optional(2020-05-15 06:05:30 +0000), isOnline: true, isInvisible: false, isBanned: false, role: StreamChatClient.User.Role.user, extraData: Optional(StreamChatClient.UserExtraData(name: Optional("Vit D1"), avatarURL: Optional(/images/college/1/6d378c06-5d98-4433-bc50-865e82aad6c1.jpeg))), devices: [], currentDevice: nil, mutedUsers: []), connectionId: "f707b5ac-5121-4b2c-921b-dbbf0f7f68ee"))

2020-05-15 06:05:31 +0000: ChatManager.swift(112) setUser(_:apiTokenProvider:completion:):state = disconnected(nil)`

These are my debug logs from completion handler:

`client.set(user: streamUser, tokenProvider: { tokenProvider in
            apiTokenProvider { token in
                tokenProvider(token?.token ?? "")
            }
        }, completion: { [weak self] result in
            guard let self = self else { return }

            DLog("self.client.connectionState = \(self.client.connectionState)")

            self.client.rx.connectionState.subscribe(onNext: { state in
                DLog("state = \(state)")
            }, onError: { error in
                DLog("error = \(error)")
            }, onCompleted: {
                DLog("completed")
            }) {
                DLog("disposed")
            }.disposed(by: self.disposeBag)

            DispatchQueue.main.async {
                do {
                    let connection = try result.get()
                    self.listenToEvents()
                    completion?(.success(connection.user))
                } catch {
                    completion?(.failure(.underlying(error)))
                }
            }
        })`

In previous user session I deliberately didn't open chat screen or started any subscriptions (commented that code out), but `client.rx.connectionState` still differs from `client.connectionState`. When I open a channel after that, nothing happens, no logs printed.

**Comment by buh** [comment]
Hi @vadubov 
We released 2.2.1 Can you check it for this issue?

**Comment by vadubov** [comment]
Unfortunately, it didn't help. I upgraded to 2.2.1, and nothing changed. Channels don't load and debug code above prints different states.

**Comment by buh** [comment]
Can you confirm  you resubscribed for events after relogin?

**Comment by vadubov** [comment]
It doesn't matter, I tried subscribing and not subscribing to events. As @KevinMiller-TheWing said (https://github.com/GetStream/stream-chat-swift/issues/274#issuecomment-631535488), channels doesn't load only if `ChannelsViewController` was initialized in previous user session.

**Comment by buh** [comment]
I think I found the issue. The fix in branch `fix-rx-reconnection`. Can you check it?
```ruby
  pod 'StreamChat', :git => 'https://github.com/GetStream/stream-chat-swift.git', :branch => 'fix-rx-reconnection'
  pod 'StreamChatCore', :git => 'https://github.com/GetStream/stream-chat-swift.git', :branch => 'fix-rx-reconnection'
  pod 'StreamChatClient', :git => 'https://github.com/GetStream/stream-chat-swift.git', :branch => 'fix-rx-reconnection'
```

**Comment by vadubov** [comment]
It works, thank you!

**Comment by buh** [comment]
Hi @vadubov 
Please can you check the branch again with the latest changes for the reconnection issue? We are going to release it.
```ruby
pod 'StreamChat', :git => 'https://github.com/GetStream/stream-chat-swift.git', :branch => 'fix-rx-reconnection'
pod 'StreamChatCore', :git => 'https://github.com/GetStream/stream-chat-swift.git', :branch => 'fix-rx-reconnection'
pod 'StreamChatClient', :git => 'https://github.com/GetStream/stream-chat-swift.git', :branch => 'fix-rx-reconnection'
```

**Comment by vadubov** [comment]
Works fine!

**Comment by buh** [comment]
v2.2.3 published. Please check [CHANGELOG](https://github.com/GetStream/stream-chat-swift/blob/master/CHANGELOG.md).

**Comment by b-onc** [comment]
Closed via #284
